pkg/cmd/runtime/plugins/qingcloud/qingcloud_runtime.go

```diff
@@ -0,0 +1,407 @@
+	"github.com/yunify/qingcloud-sdk-go/service"
+	"github.com/yunify/qingcloud-sdk-go/utils"
+
+	"openpitrix.io/openpitrix/pkg/cmd/runtime"
+	"openpitrix.io/openpitrix/pkg/logger"
+)
+
+func init() {
+	runtime.RegisterRuntime(new(QingcloudRuntime))
+}
+
+type QingcloudRuntime struct {
+}
+
+func (p *QingcloudRuntime) InitService() (qingcloudService *service.QingCloudService, err error) {
+	userConf, err := config.NewDefault()
+	if err != nil {
+		return
+	}
+	err = userConf.LoadUserConfig()
+	if err != nil {
+		return
+	}
+	qingcloudService, err = service.Init(userConf)
+	if err != nil {
+		return
+	}
+	return
+}
+
+func (p *QingcloudRuntime) InitClusterService() (clusterService *service.ClusterService, err error) {
+	qingcloudService, err := p.InitService()
+	if err != nil {
+		logger.Errorf("Failed to init qingcloud api service: %v", err)
+		return
+	}
+	clusterService, err = qingcloudService.Cluster(qingcloudService.Config.Zone)
+	return
+}
+
+func (p *QingcloudRuntime) InitJobService() (jobService *service.JobService, err error) {
+	qingcloudService, err := p.InitService()
+	if err != nil {
+		logger.Errorf("Failed to init qingcloud api service: %v", err)
+		return
+	}
+	jobService, err = qingcloudService.Job(qingcloudService.Config.Zone)
+	return
+}
+
+func (p *QingcloudRuntime) WaitJobs(jobIds []string, timeout time.Duration, waitInterval time.Duration) error {
+	jobService, err := p.InitJobService()
+	if err != nil {
+		logger.Errorf("Failed to init job service when wait cluster jobs %s: %v", jobIds, err)
+		return err
+	}
+	var wg sync.WaitGroup
+	wg.Add(len(jobIds))
+	done := make(chan error, 1)
+
+	for _, jobId := range jobIds {
+		go func() {
+			defer wg.Done()
+			done <- client.WaitJob(jobService, jobId, timeout, waitInterval)
+
+			waitErr, ok := <-done
+			if !ok {
+				return
+			}
+			if waitErr != nil {
+				logger.Errorf("Failed to wait cluster job [%s]: %v", jobId, waitErr)
+				err = waitErr
+				close(done)
+			}
+		}()
+	}
+	wg.Wait()
+	return err
+}
+
+func (p *QingcloudRuntime) WaitClustersStatus(clusterService *service.ClusterService, clusterIds string, status string, timeout time.Duration, waitInterval time.Duration) error {
+	return utils.WaitForSpecificOrError(func() (bool, error) {
+		clusters := strings.Split(clusterIds, ",")
+		output, err := clusterService.DescribeClusters(&service.DescribeClustersInput{Clusters: service.StringSlice(clusters)})
+		if err != nil {
+			//network or api error
+			return false, nil
+		}
+		if len(output.ClusterSet) != len(clusters) {
+			return false, fmt.Errorf("Can not find clusters [%s]", clusterIds)
+		} else {
+			for _, cluster := range output.ClusterSet {
+				if cluster.TransitionStatus == nil {
+					logger.Errorf("Cluster [%s] transition_status is nil ", clusterIds)
+					return false, nil
+				}
+				if service.StringValue(cluster.TransitionStatus) != "" {
+					return false, nil
+				}
+				if service.StringValue(cluster.Status) == status {
+					continue
```

Comment:
logic is complex

Reply:
wait the cluster till status transform to the defined status, if transition_status is not "" means the cluster is working,

Reply:
```go
func (p *QingcloudRuntime) WaitClustersStatus(clusterService *service.ClusterService, clusterIds, status string, timeout time.Duration, waitInterval time.Duration) error {
	return utils.WaitForSpecificOrError(func() (bool, error) {
		clusters := strings.Split(clusterIds, ",")
		output, err := clusterService.DescribeClusters(&service.DescribeClustersInput{Clusters: service.StringSlice(clusters)})
		if err != nil {
			//network or api error
			return false, nil
		}

		if len(output.ClusterSet) != len(clusters) {
			return false, fmt.Errorf("Can not find clusters [%s]", clusterIds)
		} 
		 
		for _, cluster := range output.ClusterSet {
			if cluster.TransitionStatus == nil {
				logger.Errorf("Cluster [%s] transition_status is nil ", clusterIds)
				return false, nil
			}
			if service.StringValue(cluster.TransitionStatus) != "" {
				return false, nil
			}

			if service.StringValue(cluster.Status) != status {
				return false, fmt.Errorf("Cluster [%s] status [%s]", clusterIds, status)
			}

			// continue
		}
		return true, nil
	
	}, timeout, waitInterval)
}
```

Reply:
ok